apps/builder/app/shared/commands-emitter.ts

```diff
@@ -0,0 +1,104 @@
+import { isHotkeyPressed } from "react-hotkeys-hook";
+import { atom, onMount } from "nanostores";
+import { $publisher, subscribe } from "~/shared/pubsub";
+
+type CommandMeta<CommandName extends string> = {
+  // @todo category, description
+  name: CommandName;
+  // default because hotkeys can be customized from ui
+  defaultHotkeys?: string[];
+};
+
+type CommandHandler = (source: string) => void;
+
+/**
+ * Command can be registered by builder, canvas or plugin
+ */
+type Command<CommandName extends string> = CommandMeta<CommandName> & {
+  /**
+   * Command handler accepting source where was triggered
+   * which is builder, canvas or plugin name
+   */
+  handler: CommandHandler;
+};
+
+/*
+ * expose command metas to synchronize between builder, canvas and plugins
+ */
+export const $commandMetas = atom(new Map<string, CommandMeta<string>>());
+
+export const createCommandsEmitter = <CommandName extends string>({
```

Comment:
IMO emitter is something this `on(cmd, handler)`  and `emit` command, here the result type is not even similar

Reply:
There is subscribe and emit. Actual handlers are moved into config. Idk dispatcher will work for you?